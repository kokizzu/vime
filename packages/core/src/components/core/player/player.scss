vime-player {
  /**
   * @prop --vime-font-family: A custom font family to be used throughout the player.
   * 
   * @prop --vime-theme: A custom theme (color) to be used throughout the player. Any valid CSS 
   * `color` property should work (HEX, RGBA, HLS, ...).
   */

  --vime-font-family: 'Helvetica Neue', 'Segoe UI', Helvetica, Arial, sans-serif;

  box-sizing: border-box;
  direction: ltr;
  font-family: var(--vime-font-family);
  -moz-osx-font-smoothing: auto;
  -webkit-font-smoothing: subpixel-antialiased;
  -webkit-tap-highlight-color: transparent;
  font-variant-numeric: tabular-nums;
  font-weight: 500;
  line-height: 1.7;
  width: 100%;
  display: block;
  max-width: 100%;
  min-width: 414px;
  position: relative;
  text-shadow: none;
  outline: 0;
  transition: box-shadow 0.3s ease;

  *,
  *::after,
  *::before {
    box-sizing: inherit;
  }

  button {
    font: inherit;
    line-height: inherit;
    width: auto;

    &::-moz-focus-inner {
      border:0;
    }
  }

  // Fix 300ms delay
  a,
  button,
  input,
  label {
    touch-action: manipulation;
  }

  &.video {
    height: 0;
    background: #000;
    overflow: hidden;
  }

  &.fullscreen {
    background: #000;
    margin: 0;
    border-radius: 0;
    width: 100%;
    height: 100%;
    padding-bottom: 0 !important;
  }

  .blocker {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: $z-index-blocker;
    display: inline-block;
  }
}
