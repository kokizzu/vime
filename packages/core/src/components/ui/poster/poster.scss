vime-poster {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: #000;
  z-index: $z-index-poster;
  display: none;
  opacity: 0;
  visibility: hidden;
  pointer-events: none;
  transition: $fade-transition;
    
  &.enabled {
    display: inline-block;
  }

  &.active {
    opacity: 1;
    visibility: visible;
  }

  & > img {
    width: 100%;
    height: 100%;
  }
}
