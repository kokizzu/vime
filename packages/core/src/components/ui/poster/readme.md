# vime-poster

Loads the poster set in the player prop `currentPoster` and displays it. The poster will automatically 
dissapear once playback starts.

## Example

```html
<vime-player>
  <!-- ... -->
  <vime-ui>
    <vime-poster></vime-poster>
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                                                                                   | Type                                                                    | Default   |
| -------- | --------- | --------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------- | --------- |
| `fit`    | `fit`     | How the poster image should be resized to fit the container (sets the `object-fit` property). | `"contain" \| "cover" \| "fill" \| "none" \| "scale-down" \| undefined` | `'cover'` |


## Events

| Event      | Description                             | Type                |
| ---------- | --------------------------------------- | ------------------- |
| `loaded`   | Emitted when the poster has loaded.     | `CustomEvent<void>` |
| `willHide` | Emitted when the poster will be hidden. | `CustomEvent<void>` |
| `willShow` | Emitted when the poster will be shown.  | `CustomEvent<void>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
