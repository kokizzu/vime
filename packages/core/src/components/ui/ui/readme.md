# vime-ui

Simple container that holds a collection of user interface components.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
