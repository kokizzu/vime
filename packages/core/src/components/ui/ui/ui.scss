vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}