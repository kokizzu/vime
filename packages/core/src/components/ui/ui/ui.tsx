import { h, Component } from '@stencil/core';

@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  render() {
    return (<slot />);
  }
}
