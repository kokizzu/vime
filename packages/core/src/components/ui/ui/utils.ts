import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
