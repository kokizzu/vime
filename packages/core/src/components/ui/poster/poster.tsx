import {
  h, Component, Prop, State, Watch, Host,
} from '@stencil/core';
import { openPlayerWormhole } from '../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { isUndefined } from '../../../utils/unit';

@Component({
  tag: 'vime-poster',
  styleUrl: 'poster.scss',
})
export class Poster {
  @State() isEnabled = false;

  @State() isActive = false;

  /**
   * How the poster image should be resized to fit the container (sets the `object-fit` property).
   */
  @Prop() fit?: 'fill' | 'contain' | 'cover' | 'scale-down' | 'none' = 'cover';

  /**
   * @internal
   */
  @Prop() isVideoView!: PlayerProps[PlayerProp.IsVideoView];

  /**
   * @internal
   */
  @Prop() currentPoster?: PlayerProps[PlayerProp.CurrentPoster];

  /**
   * @internal
   */
  @Prop() mediaTitle?: PlayerProps[PlayerProp.MediaTitle];

  /**
   * @internal
   */
  @Prop() playbackStarted!: PlayerProps[PlayerProp.PlaybackStarted];

  @Watch('isVideoView')
  @Watch('currentPoster')
  onEnabledChange() {
    this.isEnabled = this.isVideoView && !isUndefined(this.currentPoster);
  }

  @Watch('playbackStarted')
  onActiveChange() {
    this.isActive = !this.playbackStarted;
  }

  render() {
    return (
      <Host
        class={{
          enabled: this.isEnabled,
          active: this.isActive,
        }}
      >
        <img
          style={{ objectFit: this.fit }}
          alt={!isUndefined(this.mediaTitle) ? `${this.mediaTitle} Poster` : 'Media Poster'}
          src={this.currentPoster}
        />
      </Host>
    );
  }
}

openPlayerWormhole(Poster, [
  PlayerProp.MediaTitle,
  PlayerProp.CurrentPoster,
  PlayerProp.PlaybackStarted,
  PlayerProp.IsVideoView,
]);
