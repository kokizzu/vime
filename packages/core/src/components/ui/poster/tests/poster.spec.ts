import { SpecPage } from '@stencil/core/testing';
import { Poster } from '../poster';
import { newUISpecPage } from '../../ui/tests';
import { ViewType } from '../../../core/player/ViewType';

let page: SpecPage;
let provider: HTMLVimeFaketubeElement;
let poster: HTMLVimePosterElement;

beforeEach(async () => {
  ({ page, provider } = await newUISpecPage(
    [Poster],
    '<vime-poster />',
  ));

  poster = page.root!.querySelector('vime-poster')!;
});

const findImage = () => page.root!.querySelector('vime-poster > img');

it('should be structurally sound', () => {
  expect(poster).toMatchSnapshot();
});

it('should not render if not a video view', async () => {
  await provider.dispatchChange('viewType', ViewType.Audio);
  await page.waitForChanges();
  expect(poster).toHaveClass('hidden');
});

it('should render if currentPoster exists', async () => {
  await provider.dispatchChange('viewType', ViewType.Video);
  await provider.dispatchChange('currentPoster', '');
  await page.waitForChanges();
  expect(poster).not.toHaveClass('hidden');
});

it('should not render if currentPoster does not exist', async () => {
  expect(poster).toHaveClass('hidden');
});

// `loaded` event on `<img>` is not triggered.
it.skip('should be visible if playback has not started', () => {
  expect(poster).toHaveClass('active');
});

it('should not be visible if playback has started', async () => {
  await provider.dispatchChange('playbackStarted', true);
  await page.waitForChanges();
  expect(poster).not.toHaveClass('active');
});

it('should set the alt text based on the media title', async () => {
  expect(findImage()).toEqualAttribute('alt', 'Media Poster');
  await provider.dispatchChange('mediaTitle', 'Apples');
  await page.waitForChanges();
  expect(findImage()).toEqualAttribute('alt', 'Apples Poster');
});

it('should emit vWillShow event when visible', async () => {
  const cb = jest.fn();
  poster.addEventListener('vWillShow', cb);
  await provider.dispatchChange('viewType', ViewType.Video);
  await provider.dispatchChange('currentPoster', '');
  await page.waitForChanges();
  expect(cb).toHaveBeenCalled();
});

it('should emit vWillHide event when not visible', async () => {
  const cb = jest.fn();
  poster.addEventListener('vWillHide', cb);
  await provider.dispatchChange('viewType', ViewType.Video);
  await provider.dispatchChange('currentPoster', '');
  await page.waitForChanges();
  await provider.dispatchChange('playbackStarted', true);
  await page.waitForChanges();
  expect(cb).toHaveBeenCalled();
});
