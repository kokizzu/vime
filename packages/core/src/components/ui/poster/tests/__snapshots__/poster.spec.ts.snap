// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-poster class="active">
  <img alt="Media Poster" style="object-fit: cover;">
</vime-poster>
`;
